File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {

        int n = coins.length;
        int[][] dp = new int[n][amount + 1];

        for(int i = 0 ; i <= amount ; i++)
            if(i % coins[0] == 0)
                dp[0][i] = 1;

        for(int i = 1 ; i < n ; i++){
            for(int sum = 0 ; sum <= amount; sum++){

                int pick = 0;

                if(coins[i] <= sum){
                    pick = dp[i][sum - coins[i]];
                }

                int not_pick = dp[i - 1][sum];

                dp[i][sum] = pick + not_pick;

            }
        }

        return dp[n - 1][amount];
    }
}

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        //Using unsigned instead of int prevents signed integer overflow 
        vector<vector<unsigned>> g(coins.size() + 1, vector<unsigned> (amount+1));
        for(int i = 0 ; i < coins.size()+1; i++){
            g[i][0] = 1;
        }
        for(int i = 1; i <= coins.size(); i++){
            for(int j = 1; j <=amount; j++){
                if(j - coins[i-1] < 0){
                    g[i][j] = g[i-1][j];
                }else{
                    g[i][j] = g[i-1][j]+ g[i][j-coins[i-1]];
                }
            }
        }
        return g[coins.size()][amount];
    }
};